sa/sa.go

```diff
@@ -1291,3 +1292,136 @@ func (ssa *SQLStorageAuthority) UpdateCRLShard(ctx context.Context, req *sapb.Up
+	identifiers, err := newIdentifierModelsFromPB(req.Identifiers)
+	if err != nil {
+		return nil, err
+	}
+
+	response := &sapb.PauseIdentifiersResponse{}
+	_, err = db.WithTransaction(ctx, ssa.dbMap, func(tx db.Executor) (interface{}, error) {
+		for _, identifier := range identifiers {
+			pauseError := func(op string, err error) error {
+				return fmt.Errorf("while %s identifier %s for registration ID %d: %w",
+					op, identifier.Value, req.RegistrationID, err,
+				)
+			}
+
+			var entry pausedModel
+			err := tx.SelectOne(ctx, &entry, `
+			SELECT pausedAt, unpausedAt
+			FROM paused
+			WHERE 
+				registrationID = ? AND 
+				identifierType = ? AND 
+				identifierValue = ?`,
+				req.RegistrationID,
+				identifier.Type,
+				identifier.Value,
+			)
+			if err != nil && !errors.Is(err, sql.ErrNoRows) {
+				return nil, pauseError("querying pause status for", err)
+			}
+
+			if entry.PausedAt != nil && entry.UnpausedAt == nil {
+				// Identifier is already paused.
+				continue
+			}
+
+			if errors.Is(err, sql.ErrNoRows) {
+				// Not currently or previously paused, insert a new pause record.
+				_, err = tx.ExecContext(ctx, `
+			INSERT INTO paused (
+				registrationID,
+				identifierType,
+				identifierValue,
+				pausedAt
+			) VALUES (?, ?, ?, ?)`,
+					req.RegistrationID,
+					identifier.Type,
+					identifier.Value,
+					ssa.clk.Now().Truncate(time.Second),
+				)
+				if err != nil && !db.IsDuplicate(err) {
+					return nil, pauseError("pausing", err)
+				}
+
+				// Identifier successfully paused.
+				response.Paused++
+				continue
+			}
+
+			if entry.PausedAt == nil && entry.UnpausedAt != nil {
+				// Previously paused (and unpaused), repause the identifier.
+				_, err := tx.ExecContext(ctx, `
+				UPDATE paused
+				SET pausedAt = ?,
+					unpausedAt = NULL
+				WHERE 
+					registrationID = ? AND 
+					identifierType = ? AND 
+					identifierValue = ? AND
+					pausedAt IS NULL AND 
+					unpausedAt IS NOT NULL`,
+					ssa.clk.Now().Truncate(time.Second),
+					req.RegistrationID,
+					identifier.Type,
+					identifier.Value,
+				)
+				if err != nil {
+					return nil, pauseError("repausing", err)
+				}
+
+				// Identifier successfully repaused.
+				response.Repaused++
+				continue
+			} else {
+				// Paused table entry is in an invalid state.
+				return nil, fmt.Errorf("identifier %s for registration ID %d is in an unexpected state",
+					identifier.Value,
+					req.RegistrationID,
+				)
+			}
+		}
+		return nil, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// UnpauseAccount will unpause all paused identifiers for the provided account.
+// If no identifiers are currently paused, this is a no-op.
+func (ssa *SQLStorageAuthority) UnpauseAccount(ctx context.Context, req *sapb.RegistrationID) (*emptypb.Empty, error) {
+	if core.IsAnyNilOrZero(req.Id) {
+		return nil, errIncompleteRequest
+	}
+
+	_, err := ssa.dbMap.ExecContext(ctx, `
```

Comment:
I _believe_ that, if zero rows are affected, this will still return successfully. That should be double-checked to ensure we're not accidentally returning an ErrNoRows or something like that.

And/or there should be a test for unpausing a never-paused account.

Reply:
ErrNoRows is only returned for SELECT statements. For ExecContext you would need to check the sql.Result response to determine if there were 0 rows impacted.